aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,786 @@
+        }
+        let outlen: usize = outlen.try_into()?;
+        Ok((self.context.into(), BufferUpdate::new(output, outlen)))
+    }
+
+    /// Returns the cipher operating mode.
+    #[must_use]
+    pub fn mode(&self) -> OperatingMode {
+        self.mode
+    }
+
+    /// Returns the cipher algorithm.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+
+    /// Constructs a `StreamingEncryptingKey` for encrypting data using the CTR cipher mode.
+    /// The resulting ciphertext will be the same length as the plaintext.
+    ///
+    /// # Errors
+    /// Returns and error on an internal failure.
+    pub fn ctr(key: UnboundCipherKey) -> Result<Self, Unspecified> {
+        let context = key.algorithm().new_encryption_context(OperatingMode::CTR)?;
+        Self::less_safe_ctr(key, context)
+    }
+
+    /// Constructs a `StreamingEncryptingKey` for encrypting data using the CTR cipher mode.
+    /// The resulting ciphertext will be the same length as the plaintext.
+    ///
+    /// This is considered less safe because the caller could potentially construct
+    /// an `EncryptionContext` from a previously used initialization vector (IV).
+    ///
+    /// # Errors
+    /// Returns an error on an internal failure.
+    pub fn less_safe_ctr(
+        key: UnboundCipherKey,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CTR, context)
+    }
+
+    /// Constructs a `StreamingEncryptingKey` for encrypting data using the CBC cipher mode
+    /// with pkcs7 padding.
+    /// The resulting ciphertext will be longer than the plaintext; padding is added
+    /// to fill the next block of ciphertext.
+    ///
+    /// # Errors
+    /// Returns an error on an internal failure.
+    pub fn cbc_pkcs7(key: UnboundCipherKey) -> Result<Self, Unspecified> {
+        let context = key.algorithm().new_encryption_context(OperatingMode::CBC)?;
+        Self::less_safe_cbc_pkcs7(key, context)
+    }
+
+    /// Constructs a `StreamingEncryptingKey` for encrypting data using the CBC cipher mode
+    /// with pkcs7 padding.
+    /// The resulting ciphertext will be longer than the plaintext; padding is added
+    /// to fill the next block of ciphertext.
+    ///
+    /// This is considered less safe because the caller could potentially construct
+    /// an `EncryptionContext` from a previously used initialization vector (IV).
+    ///
+    /// # Errors
+    /// Returns an error on an internal failure.
+    pub fn less_safe_cbc_pkcs7(
+        key: UnboundCipherKey,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CBC, context)
+    }
+}
+
+/// A key for streaming decryption operations.
+pub struct StreamingDecryptingKey {
+    algorithm: &'static Algorithm,
+    mode: OperatingMode,
+    cipher_ctx: LcPtr<EVP_CIPHER_CTX>,
+}
+impl StreamingDecryptingKey {
```

Comment:
Just leaving a comment reminder to determine whether we should assert the `outlen` written check here as indicated above for update and finish on this type.